modules/junit-jupiter/src/main/java/org/testcontainers/junit/jupiter/TestcontainersExtension.java

```diff
@@ -0,0 +1,125 @@
+    public void beforeEach(final ExtensionContext context) {
+        collectParentTestInstances(context)
+            .stream()
+            .flatMap(this::findRestartedContainers)
+            .forEach(adapter -> context.getStore(NAMESPACE)
+                .getOrComputeIfAbsent(adapter.key, k -> adapter.start()));
+    }
+
+    private Set<Object> collectParentTestInstances(final ExtensionContext context) {
+        Set<Object> testInstances = new LinkedHashSet<>();
+        Optional<ExtensionContext> current = Optional.of(context);
+        while(current.isPresent()) {
+            ExtensionContext ctx = current.get();
+            Object testInstance = ctx.getStore(NAMESPACE).remove(TEST_INSTANCE);
+            if (testInstance != null) {
+                testInstances.add(testInstance);
+            }
+            current = ctx.getParent();
+        }
+        return testInstances;
+    }
+
+    private Stream<StoreAdapter> findSharedContainers(Object testInstance) {
+        return ReflectionUtils.findFields(
+                testInstance.getClass(),
+                isSharedContainer(),
+                ReflectionUtils.HierarchyTraversalMode.TOP_DOWN)
+            .stream()
+            .map(f -> getContainerInstance(testInstance, f));
+    }
+
+    private Predicate<Field> isSharedContainer() {
+        return isContainer().and(ReflectionUtils::isStatic);
+    }
+
+    private Stream<StoreAdapter> findRestartedContainers(Object testInstance) {
+        return ReflectionUtils.findFields(
+                testInstance.getClass(),
+                isRestartContainer(),
+                ReflectionUtils.HierarchyTraversalMode.TOP_DOWN)
+            .stream()
+            .map(f -> getContainerInstance(testInstance, f));
+    }
+
+    private Predicate<Field> isRestartContainer() {
+        return isContainer().and(ReflectionUtils::isNotStatic);
+    }
+
+    private static Predicate<Field> isContainer() {
+        return field -> Startable.class.isAssignableFrom(field.getType()) && AnnotationSupport.isAnnotated(field, Container.class);
+    }
+
+    private static StoreAdapter getContainerInstance(final Object testInstance, final Field field) {
+        try {
+            field.setAccessible(true);
+            Startable containerInstance = Preconditions.notNull((Startable) field.get(testInstance), "Container " + field.getName() + " needs to be initialized");
+            return new StoreAdapter(field.getDeclaringClass(), field.getName(), containerInstance);
+        } catch (IllegalAccessException e) {
+            throw new ExtensionConfigurationException("Can not access container defined in field " + field.getName());
+        }
+    }
+
+    /**
+     * An adapter for {@link Startable} that implement {@link CloseableResource}
+     * thereby letting the JUnit automatically stop containers once the current
+     * {@link ExtensionContext} is closed.
+     */
+    private static class StoreAdapter implements CloseableResource {
```

Comment:
Unfortunately no. It has to implement `ClosableResource` from JUnit in order to be closed automatically.

Reply:
It would have been a breaking change if out of sudden all stored `AutoCloseable` instances were closed when the context of the store reached its end of life. Therefore, we introduced that flagging interface.

Reply:
Thanks for the clarification.